x-pack/plugins/cases/public/components/all_cases/use_all_cases_query_params.tsx

```diff
@@ -0,0 +1,127 @@
+ * 2.0.
+ */
+
+import { useCallback, useRef, useState } from 'react';
+import { useLocation, useHistory } from 'react-router-dom';
+import { isEqual } from 'lodash';
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { parse, stringify } from 'query-string';
+
+import { DEFAULT_QUERY_PARAMS } from '../../containers/use_get_cases';
+import { parseUrlQueryParams } from './utils';
+import { LOCAL_STORAGE_KEYS } from '../../../common/constants';
+
+import type {
+  LocalStorageQueryParams,
+  ParsedUrlQueryParams,
+  QueryParams,
+  UrlQueryParams,
+} from '../../../common/ui/types';
+import { useCasesContext } from '../cases_context/use_cases_context';
+
+export const getQueryParamsLocalStorageKey = (appId: string) => {
+  const filteringKey = LOCAL_STORAGE_KEYS.casesFiltering;
+  return `${appId}.${filteringKey}`;
+};
+
+const getQueryParams = (
+  params: UrlQueryParams,
+  queryParams: UrlQueryParams,
+  urlParams: UrlQueryParams,
+  localStorageQueryParams?: LocalStorageQueryParams
+): QueryParams => {
+  const result = { ...DEFAULT_QUERY_PARAMS };
+
+  result.perPage =
+    params.perPage ??
+    urlParams.perPage ??
+    localStorageQueryParams?.perPage ??
+    DEFAULT_QUERY_PARAMS.perPage;
+
+  result.sortField = params.sortField ?? queryParams.sortField ?? DEFAULT_QUERY_PARAMS.sortField;
+
+  result.sortOrder =
+    params.sortOrder ??
+    urlParams.sortOrder ??
+    localStorageQueryParams?.sortOrder ??
+    DEFAULT_QUERY_PARAMS.sortOrder;
+
+  result.page = params.page ?? urlParams.page ?? DEFAULT_QUERY_PARAMS.page;
+
+  return result;
+};
+
+export function useAllCasesQueryParams(isModalView: boolean = false) {
+  const { appId } = useCasesContext();
+  const location = useLocation();
+  const history = useHistory();
+  const isFirstRenderRef = useRef(true);
+
+  const [queryParams, setQueryParams] = useState<QueryParams>({ ...DEFAULT_QUERY_PARAMS });
+
+  const [localStorageQueryParams, setLocalStorageQueryParams] =
+    useLocalStorage<LocalStorageQueryParams>(getQueryParamsLocalStorageKey(appId));
+
+  const persistAndUpdateQueryParams = useCallback(
+    (params) => {
+      if (isModalView) {
+        setQueryParams((prevParams) => ({ ...prevParams, ...params }));
+        return;
+      }
+
+      const parsedUrlParams: ParsedUrlQueryParams = parse(location.search);
+      const urlParams: UrlQueryParams = parseUrlQueryParams(parsedUrlParams);
+      const newQueryParams: QueryParams = getQueryParams(
+        params,
+        queryParams,
+        urlParams,
+        localStorageQueryParams
+      );
+      const newLocalStorageQueryParams = {
+        perPage: newQueryParams.perPage,
+        sortOrder: newQueryParams.sortOrder,
+      };
+      const newUrlParams = {
+        page: newQueryParams.page,
+        ...newLocalStorageQueryParams,
+      };
+
+      if (!isEqual(newUrlParams, urlParams)) {
+        try {
+          history.push({
```

Comment:
Cases are being used in Observability, Security solution, and the Stack. We decided it is better to push the history so the user can go back to the previous filtering. Why did you decide to not keep the history in your queries?

Reply:
Hi @machadoum , thanks for the review!

>  I noticed that the cases page has sourcerer and timerange query params. Are they used for anything? Does it make sense to delete it?

These are actually not ours, they come from your side and I tried to leave everything as is. In fact, at some point, the logic did not take them into account and the security tests failed.

Reply:
That decision preceded me. But I can imagine that one reason is that security has many filters and it could take many clicks for a user to return to the previous page.
Especially if you consider that filters are added the first time a page loads. So if a user opens the cases page and decides to immediately go to the previous page, it wouldn't work. They would have to click at least twice. I tested some pages and it is very inconsistent across applications. For example, on the Observability Alerts page, the user has to click 4 times the back button to leave the page.

Reply:
@mdefazio Any input on this?

Reply:
> Unrelated question: I noticed that the cases page has sourcerer and timerange query params. Are they used for anything? Does it make sense to delete it?

@machadoum I think this params are used by the timeline, the TL bottom bar is enabled in Cases (in Security)

Reply:
Since the back button behaviour is inconsistent across different pages and areas, let's not block this PR because of it. The only scenario that looks like a bug to me is the back button behaviour when the page loads.
Given the following scenario: 
1. The user opens the menu
2. The user clicks on the cases link
3. Cases page loads
4. The user clicks the browser back button
**Expected behaviour:**  It should navigate back to the previous page

Reply:
As discussed offline, I will merge this PR now to make FF [and created an issue for this bug behavior](https://github.com/elastic/kibana/issues/145110) .

I will handle this issue during the bug-fixing phase sometime this week.

Thanks for the review @machadoum 🙌

Reply:
I Agree, let's not block it.
But the workflow described by @machadoum is weird. I think we should discuss this issue, the back button behavior should be consistent across the app. 
(IMHO, I don't think the expected behavior of the browser's back button is to roll back the filters.)
cc @YulNaumenko